File: LeetCode/0006. ZigZag Conversion/solution.cpp
/**
 * 1157 / 1157 test cases passed.
 * Runtime: 28 ms
 * Memory Usage: 49.3 MB 
 */
class Solution {
public:
    string convert(string s, int numRows) {
        int n = s.size(), r = numRows;
        if (r == 1 || r >= n) return s;

        vector<string> mat(r);
        for (int i = 0, x = 0, period = r * 2 - 2; i < n; ++ i) {
            mat[x] += s[i];
            i % period < r - 1 ? ++ x : -- x;
        }
        return accumulate(mat.begin(), mat.end(), string(""));
    }
};

File: LeetCode/0006. ZigZag Conversion/solution.py
"""
1157 / 1157 test cases passed.
Runtime: 52 ms
Memory Usage: 15.1 MB
"""
class Solution:
    def convert(self, s: str, numRows: int) -> str:
        n, r = len(s), numRows
        if r == 1 or r >= n:
            return s
        x, period = 0, 2 * r - 2
        mat = [""] * r
        for i in range(n):
            mat[x] += s[i]
            if i % period < r - 1:
                x += 1
            else:
                x -= 1
        return "".join(mat)
